Rolling co-lead roles for T-compiler (#3262)
rust-lang/rfcs: text/0000-compiler-team-rolling-leads.md

with no concerns" process.

The leads own making decisions about urgent issues or ones with a specific
deadline. For example, deciding what to do about a critical release-blocking bug
should happen before the release, preferably long before. Likewise,
beta-backport decisions need to be made in time for the backport to happen
before the beta is lifted to stable.

Finally, when adverse events happen, the leads are responsible for reviewing
what decisions or processes led to the event, and taking action to prevent
future occurrences of the same event. Examples of this include the incr-comp bug
that plagued the 1.52.0 release, which led to a 1.52.1 release four days later
(and three steering meetings as follow-up).

## Time Commitment, Expectations, and Competencies

Let us recall that any compiler team member is allowed to:

* drive progress on backlogged work,
* draft steering meeting proposals, and often write the associated steering meeting document to drive the meetings,
* solicit individuals to form working groups to address important problems,
* take on the resolution of unassigned or abandoned P-critical or P-high issues, and

> * drive larger initiatives related to the language.

Author: I agree that compiler work should be explicitly indicated here, since there are plenty of things that are entirely compiler-internal that we would want spelled out as large initiative.

But the whole point of this section, in many ways, is to spell out things that someone might *think* is part of the role of the T-compiler leader, and yet in fact **isn't** -- its rather the role of any T-compiler member. (Arguably its the role of any project contributor who wants to stand up and do the work!)

> * drive larger initiatives related to the language.

Author: Changed my mind, went with Wesley's original suggestion, see comment on commit here for more details: https://github.com/rust-lang/rfcs/pull/3262/commits/e4f3c57fced4553e9e7c9acc4da873d5ef3b3b9a

```
enum Level { Senior, Junior }
struct Member { lead: Option<Level>, ... }

fn roll(curr_senior: &mut Member, curr_junior: &mut Member, incoming: &mut Member) {
    assert_eq!(curr_senior.lead, Some(Level::Senior));
    assert_eq!(curr_junior.lead, Some(Level::Junior));
    assert_ne!(curr_senior, incoming);
    assert_ne!(curr_junior, incoming);
    curr_senior.lead = None;
    curr_junior.lead = Some(Level::Senior);
    incoming.lead = Some(Level::Junior);
}
```

### Selection process

When the senior co-lead decides that they are ready to step down, and have
confirmed that the junior co-lead feels ready to take on the senior co-lead
role, then the two tell the T-compiler team privately about the intention to
have a rollover in leadership.

Then the T-compiler team members can nominate their teammates to serve as the
new junior co-lead. We here follow the model of our FCP process: The leads
should provide a ten-day window for nominations to come in, unless they get
confirmation that the set of nominations is complete.

After nomination is completed, the outgoing senior and junior co-leads discuss
the set of nominees, and also, if desired, have short discussions with the

> nominnees. Then, the senior and junior co-leads select the new co-lead from the

Reviewer: ```suggestion
nominees. Then, the senior and junior co-leads select the new co-lead from the
```

set of nominees. And that’s it! Then current leads just need to publish a blog
post saying that the leadership is scheduled to roll over, who the new junior
lead is, and the date that it takes effect.

# Drawbacks
[drawbacks]: #drawbacks

Why should we *not* do this?

Committing to specific term lengths puts pressure on the leads to identify new
leaders earlier than they might otherwise.

> 

Reviewer 2: Relatedly, building up leaders (especially from a finite group where trust is built over time) is a non-trivial process, and not guaranteed to work. There are several reasons that a future leader might be hard to find:
* Joining the compiler team is not pre-requisite on committing to one day take on a leadership role.
* Leadership roles are not for everyone, and that is a feature not a bug.
* Compiler team membership does not change that quickly meaning the pool of possible leaders does not change quickly.

I think it's reasonable to think that there may be situations in the future when no junior candidate is obvious. What happens in that situation?

Author: I agree we cannot force every team member to take on the leadership role.

I think the answer here is that the leadership, both jr and sr, need to be mentoring their teammates, and fellow community members, as time passes, in order to ensure that we *do* have a pool of canididates when the time comes.

You're correct that the next generation of leaders won't pop up magically.

Perhaps the most important piece here is this: While we cannot ensure that we always have a member of the compiler team who is ready to pick up the mantle, I think it is safe to say that we should always have some team member **or contributor** who seems well-suited to a leadership role. Thus, part of the job of the T-compiler leads is to idenitfy such talent, and if it includes contributors who are not yet a full team member, then the T-compiler leads should work to figure out what gaps in the experience of those contributors need to be filled before we would be comfortable promoting them to full team members.

The process I am alluding to above is indeed not a short process at all. It may be worthwhile to spell it out somewhere, though I'm not sure if this RFC is itself the right place for it...

Author: You asked a more specific question than what I addressed above, let me try again.

> I think it's reasonable to think that there may be situations in the future when no junior candidate is obvious. What happens in that situation?

I think if we truly hit a point where no junior candidate is present, then we will have hit a situation where we will need to ask the current leadership to stay on board for longer than expected. But part of the leads' goals at that point *has* to be proactive seeking of the next generation of leaders.

Asking people to stay in the role of leader ad infinitum because no one else is presenting themselves is unhealthy and not sustainable. So there needs to be an active resposne to the scenario that you describe.

Author: Hopefully the text added in 23bf9947662b306a1f4675b8e0529ed42d779a8a addresses the point raised here.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Rolling, not Rotation

We specify here a rolling leadership process, where we will see a shift in
leads, but not every team member is required to serve as a leader at any point.
An alternative model is a true *rotating leadership*, where every team member
will eventually become a lead.

Our rationale for this is that we do not think every team member is interested
in becoming a lead. If we adopted a true rotation and forced someone to become a
lead who did not really want the role, then that probably be bad for that
individual, and probably would be bad for the team as well.

## Selection, Not Election

We specify here that the new junior lead is *selected* (from a set of
individuals nominated by the compiler team) by the current leads. An obvious
alternative would be a pure democracy where the electorate (either the compiler
team, or some superset thereof) gets to vote for who the new junior lead will
be.

When it comes to co-leadership, the two leaders need to be able to work
together effectively. We do not want to end up with a pair of leaders who cannot

> talk to each other. Therefore, we currently are choosing a system where the

Reviewer 2: > a pair of leaders who cannot talk to each other

The point stands, but having two compiler team members not able to talk to each other is already a red flag. The working relationship of the two co-leads is already a consideration even if they are capable of talking to each other 😉. Maybe it would be a good idea to soften the language here to indicate that the skills and working styles of the two co-leads need to fit well together.

Author: Yes, you're right; I used imperfect language here and will soften it.

Author: Hopefully the text added in 23bf9947662b306a1f4675b8e0529ed42d779a8a addresses the point raised here.

> When it comes to co-leadership, the two leaders need to be able to work
> together effectively. We do not want to end up with a pair of leaders who cannot
> talk to each other. Therefore, we currently are choosing a system where the

Author: ```suggestion
When it comes to co-leadership, the two leaders need to be able to work
together effectively; we believe they need compatible working styles
and complementary sets of skills. Therefore, we currently are choosing a
system where the
```

    incoming.lead = Some(Level::Junior);
}
```

### Selection process

When the senior co-lead decides that they are ready to step down, and have
confirmed that the junior co-lead feels ready to take on the senior co-lead
role, then the two tell the T-compiler team privately about the intention to
have a rollover in leadership.

Then the T-compiler team members can nominate their teammates to serve as the
new junior co-lead. We here follow the model of our FCP process: The leads
should provide a ten-day window for nominations to come in, unless they get
confirmation that the set of nominations is complete.

After nomination is completed, the outgoing senior and junior co-leads discuss
the set of nominees, and also, if desired, have short discussions with the
nominees. Then, the senior and junior co-leads select the new co-lead from the
set of nominees. And that’s it! Then current leads just need to publish a blog
post saying that the leadership is scheduled to roll over, who the new junior
lead is, and the date that it takes effect.

# Drawbacks
[drawbacks]: #drawbacks

Why should we *not* do this?

Committing to specific term lengths puts pressure on the leads to identify new

> leaders earlier than they might otherwise. Note that If the leads fail to identify 

Author: ```suggestion
leaders earlier than they might otherwise. Note that if the leads fail to identify 
```
